### PR TITLE
Create ability system

### PR DESCRIPTION
**Pull request by CraZyB1336**
Ability system is now finished.

To use it, all nodes require a node with the script AbilityHandler.

Each ability except for the main abilities are resources containing an icon and an animation.

Main abilities are found by using node paths and the name for the methods as a string.
image

Each resource is a delegate where methods can subscribe to.

### CODE REVIEW

File: scripts/classes/soldier/abilities/SoldierPrimary.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+		// int count = 0;
+		// Node root = GetTree().Root;
+		// foreach(Node child in root.GetChildren())
+		// {
+		// 	if (child is BulletHole)
+		// 		count++;
+		// }
+
+		// GD.Print("BulletHole scenes: ", count);
+	}
```
Comment (ChrisKrane) [inline scripts/classes/soldier/abilities/SoldierPrimary.cs]: Debugging code still needed?

File: scripts/classes/soldier/abilities/SoldierPrimary.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+		if (!hasShot)
+		{
+			isShooting = true;
+			animationPlayer.Play("GunShot");
+			
+			Node3D bullet = (Node3D) bulletScene.Instantiate();
+			bullet.GlobalTransform = bulletSpawnpoint.GlobalTransform;
+			Node root = GetTree().Root;
+			root.AddChild(bullet);
+			hasShot = true;
```
Comment (ChrisKrane) [inline scripts/classes/soldier/abilities/SoldierPrimary.cs]: Would like some docs for summary of functions, especially for important functions like shoot

**Review by ChrisKrane — APPROVED** [review]
Comments etc is not critical, but I would like to see comments at some point 💬

File: scripts/player/Ability.cs [hunk 3]
```diff
@@ -2,21 +2,27 @@
...
 public partial class Ability : Resource
 {
+	// This is the texture for the icon in the UI.
 	[Export]
 	public Texture2D icon { get; set; }
 
+	// This is the animation the player will play when using the ability.
+	// Should we have this here? Can be refactored.
 	[Export]
 	public Animation playerAnimation { get; set; }
```
Comment (ChrisKrane) [inline scripts/player/Ability.cs:13]: Refactor how?

File: scripts/player/BaseCharacter.cs [hunk 4]
```diff
@@ -28,24 +28,31 @@ public override void _Ready()
 
 	public override void _PhysicsProcess(double delta)
 	{
+		// We retrieve the current velocity, since it is readOnly, and thus create a new "instance" of it.
 		Vector3 velocity = Velocity;
+
+		// Jump handling. IsOnFloor() is a built in function for testing wether. 
 		if (IsOnFloor() && Input.IsActionJustPressed("Jump"))
 		{
```
Comment (ChrisKrane) [inline scripts/player/BaseCharacter.cs]: "Testing wether."

File: scripts/player/BaseCharacter.cs [hunk 5]
```diff
@@ -78,6 +86,7 @@ private Vector3 HandleAirVelocity(Vector3 wishDirection, Vector3 currentVelocity
 		float currentSpeed = currentVelocity.Dot(wishDirection);
 
 		// Arbitrary values, can be changed or customized if necessary.
+		// TBH idk how this works, look up the quake air strafe movement.
 		float addedSpeed = Mathf.Max(2*speed/3 - currentSpeed, 0);
```
Comment (ChrisKrane) [inline scripts/player/BaseCharacter.cs:90]: Based.

**Review by ChrisKrane — APPROVED** [review]
Non-critical comments only. Good stuff

**Comment by CraZyB1336** [comment]
👉👈🥺

File: scripts/player/Ability.cs (hunk 3 above)
Comment (CraZyB1336) [reply scripts/player/Ability.cs:13]: Refactored in either removed or a different implementation

File: scripts/player/BaseCharacter.cs (hunk 4 above)
Comment (CraZyB1336) [reply scripts/player/BaseCharacter.cs]: woops

**Review by ChrisKrane — APPROVED** [review]
🐐

**Pull request merged**